src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,130 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="state">The state associated with the target history entry.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation. If the result is <see langword="true"/>, the navigation should continue.</returns>
+    protected async ValueTask<bool> NotifyLocationChangingAsync(string uri, string? state, bool isNavigationIntercepted)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        var handlerCount = _locationChangingHandlers.Count;
+
+        if (handlerCount == 0)
+        {
+            return true;
+        }
+
+        _locationChangingCts = new();
+
+        var context = new LocationChangingContext(uri, state, isNavigationIntercepted, _locationChangingCts);
+        var cancellationToken = _locationChangingCts.Token;
+
+        try
+        {
+            if (handlerCount == 1)
+            {
+                var handlerTask = _locationChangingHandlers[0].Invoke(context);
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return false;
+                }
+
+                if (!handlerTask.IsCompletedSuccessfully)
+                {
+                    await handlerTask.AsTask().WaitAsync(cancellationToken);
+                }
+            }
+            else
+            {
+                var locationChangingHandlersCopy = ArrayPool<LocationChangingHandler>.Shared.Rent(handlerCount);
+
+                try
+                {
+                    _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+                    var locationChangingTasks = new Task[handlerCount];
+
+                    for (var i = 0; i < handlerCount; i++)
+                    {
+                        var handlerTask = locationChangingHandlersCopy[i].Invoke(context);
+
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            return false;
+                        }
+
+                        locationChangingTasks[i] = handlerTask.AsTask();
```

Comment:
I don't think using `ValueTask` here adds a lot of value over using `Task` directly. There is no return value so a handler can do return Task.CompletedTask if their handler executes synchronously and we are having to do `AsTask()` in all cases but the simplest one.